Add edit button to readme

paulmolluzzo opened the pull request:
Much of this was taken from @devkhan's work, I just tweaked it a bit.

This takes a file that has something like `readme` or `readme.*` at the of the file name and uses that to make an edit link, then shoves it in the upper left corner on hover on the main README.

-

Fixes #40.

sindresorhus commented on extension/content.js:
```diff
@@ -170,6 +170,27 @@ function addBlameParentLinks() {
+		}
+	});
+
+	const editButtonHtml = '<div id="refined-readme-edit-link">' +
```
use template literal

sindresorhus commented on extension/content.css:
```diff
@@ -163,3 +163,11 @@
+
+/* style for edit README button */
+
+#readme:hover #refined-readme-edit-link {
```
```
#refined-github-readme-edit-link
```

sindresorhus commented on extension/content.css:
```diff
@@ -163,3 +163,11 @@
 }
+
+/* style for edit README button */
+
```
remove this empty line

sindresorhus commented on extension/content.js:
```diff
@@ -170,6 +170,27 @@ function addBlameParentLinks() {
 	});
 }
 
+function addEditButton() {
```
addReadmeEditButton

sindresorhus commented:
Hmm, it's kinda distracting that the button appear whenever you hover over the readme. Maybe we should try just putting the pencil icon (no text) permanently at the top-right corner of the readme, faded out, and when you hover over the pencil it get's 100% opacity.

sindresorhus commented on extension/content.js:
```diff
@@ -194,6 +215,8 @@ document.addEventListener('DOMContentLoaded', () => {
 	}
 
 	if (isRepo) {
+		const isRepoTree = window.location.href.split('/').length === 5 || window.location.href.split('/')[5] === 'tree';
```
I don't think we should show it in `window.location.href.split('/')[5] === 'tree'` and I would call it `isRepoRoot` instead.

sindresorhus replied:
Also drop the `window` prefix

paulmolluzzo replied:
removing the test for `window.location.href.split('/')[5] === 'tree'` means you can't edit in non-default branches. E.g. you can't edit stuff in the `development` branch. I think we should leave it, no?

sindresorhus replied:
Yeah. I thought it was a leftover. Leave it.

devkhan replied:
Is ```isRepoRoot``` ok to check? As it will only only allow to edit the main readme. There are readmes in sub-directories also.

sindresorhus replied:
We're only interested in the top readme in any branch.

paulmolluzzo replied:
The _real_ root is `location.pathname === `/${ownerName}/${repoName}/``. That's making sure you're at `/sindresorhus/refined-github/` for example. If you want to support `/sindresorhus/refined-github/tree/master` you have to support all trees since it's possible that the default branch is not master (e.g. having `gh-pages` as the default for a jekyll site or something).

sindresorhus replied:
Yes

sindresorhus commented on extension/content.js:
```diff
@@ -170,6 +170,27 @@ function addBlameParentLinks() {
+		if (isReadme) {
+			editHref = el.href.replace('blob', 'edit');
+		}
+	});
```
You don't need all this. You can get the readme name from `$('#readme > h3').text().trim()`. Then just construct the URL manually.

paulmolluzzo replied:
The issue with that is I have to grab the branch name - `master` is not always the current branch being viewed. What do you think?

sindresorhus replied:
I still think that would be cleaner. You can easily get the master branch `$('#branch').attr('value')`.

paulmolluzzo replied:
LOL, how did you find that?

sindresorhus replied:
View source, Cmd+F, `master`.

paulmolluzzo replied:
I realized it after I commented...

sindresorhus commented on extension/content.js:
```diff
@@ -170,6 +170,20 @@ function addBlameParentLinks() {
+														`<a href="${editHref}">` +
+															`<svg class="octicon octicon-pencil" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path d="M0 12v3h3l8-8-3-3L0 12z m3 2H1V12h1v1h1v1z m10.3-9.3l-1.3 1.3-3-3 1.3-1.3c0.39-0.39 1.02-0.39 1.41 0l1.59 1.59c0.39 0.39 0.39 1.02 0 1.41z"></path></svg>` +
+														`</a>` +
+													`</div>`;
```
You don't need `+` with template literal. Just put everything into one template literal

sindresorhus commented on extension/content.css:
```diff
@@ -163,3 +163,16 @@
+	position: absolute;
+	top: 10px;
+	right: 10px;
+	opacity: 0;
```
Maybe 0.2?

sindresorhus commented on extension/content.css:
```diff
@@ -163,3 +163,16 @@
+	transition: opacity 250ms;
+}
+
+#readme:hover #refined-github-readme-edit-link {
```
I think it should fade in when you hover over the pencil, not the readme.

paulmolluzzo replied:
Nice. Also like the 0.2 default better. Reduced the specificity, no need for the `#readme` selector since it's already an ID.

sindresorhus commented on extension/content.js:
```diff
@@ -194,6 +208,8 @@ document.addEventListener('DOMContentLoaded', () => {
 	}
 
 	if (isRepo) {
+		const isRepoRoot = location.pathname === `/${ownerName}/${repoName}/`;
```
The root is usually without the last `/`, but should handle both cases.

paulmolluzzo replied:
OK, handling both cases now in the `isRepoRoot` test and also in the construction of the `editHref`.

paulmolluzzo commented:
@sindresorhus should be ready to go.

sindresorhus commented:
Could you also add it to the feature list in the readme?

sindresorhus commented:
Seems to not be working on alternative branches: https://github.com/sindresorhus/ava/tree/0.9.x

paulmolluzzo commented:
OK, fixed what @sindresorhus mentioned. The tree stuff required fixing the `editHref` and the place we grab the branch name (that's not on all pages I guess). It's a little verbose because I have to create the entire URL.

In the README I dropped a little line about this feature and added a screenshot.

sindresorhus commented:
Great work @paulmolluzzo. I'm gonna use this so much :)

paulmolluzzo commented:
Thanks!
